Add silabs series 2 IADC driver with scan conversion and DMA support.

Martinhoff-maker opened the pull request:
This pull request introduces a new Silicon Labs IADC (Incremental ADC) driver for Series 2 SoCs, replaces the old device tree compatible string, and adds support for DMA-based ADC conversions. It also updates board overlays, device tree bindings, and test configurations to support and validate the new driver on multiple series 2 SoCs.

Tested all these tests and samples: adc_dt, adc_sequence, adc_api (normal, with DMA, with DMA async), adc_accuracy, adc_error_case
on all these boards:
- slwrb4180a   (xg21 SoC)
- sltb010a     (xg22 SoC) 
- xg23_rb4210a (xg23 SoC)
- xg24_rb4187c (xg24 SoC)
- xg27_dk2602a (xg27 SoC)
- xg29_rb4412a (xg29 SoC)

Also tested the adc_sequence sample with PM over xg29_rb4412 and we got an average current consumption of 13 uA.

A few ideas for improvement:
 - Implement runtime calibration (described in Silicon Labs reference manual)
 - Implement internal IADC timer instead of using the kernel timer
 - Add more possibilities for the acquisition time (which is different from interval between samplings)

FYI:
The resolution property for an ADC channel node is not mandatory. When it is not present, we set the property to 0 in the channel struct.
I initially used this "0" value to let the driver set the default resolution value. However, in the adc_error_case test case, a resolution of "0" is considered an error.
Shouldn't we set the resolution as a mandatory property since 0 is not a valid resolution value?

Martinhoff-maker committed:
drivers: adc: add IADC support for silabs series 2 boards

This commit introduces a new driver for the silabs Incremental ADC (IADC).

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
drivers: adc: add DMA support for Silabs IADC

This commit introduces DMA support for the Silabs IADC driver.
A new Kconfig option is added to enable DMA support, ensuring
compatibility with the existing ADC configuration.

DMA can be used for synch/asynch operation.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
tests: adc: add configuration and overlay for DMA support on Silabs IADC

This commit introduces a new configuration file for DMA support specific
to the Silabs IADC, along with updates to the test case YAML and device
tree overlay for the xg29_rb4412a board. The changes ensure that the ADC
driver can utilize DMA operations effectively.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
samples/tests: adc: silabs: add multiple device overlays.

This commit introduces/updates overlays for multiple silabs boards
for ADC samples and ADC tests.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
dts: arm: silabs: update IADC driver binding for all series 2 board

Update compatibility from "silabs,gecko-iadc" to "silabs,iadc". It
allows to use the new driver which are more custom to series 2 boards.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
drivers: adc: enhance millivolt conversion with 64-bit support

This commit updates the adc_raw_to_millivolts function to use 64-bit
arithmetic. It is necessary when the intermediate result is higher
than the maximum value of uint32_t. It also adds a range check to
ensure the conversion result does not exceed the limits of int32_t,
providing an assertion message for out-of-range values.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
samples: adc_sequence: add oversampling configuration option

This commit introduces a new Kconfig option for oversampling
in the ADC sequence sample. It allows to test oversampling for
IADC sequences.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker commented on drivers/adc/adc_silabs_iadc.c:
```diff
@@ -0,0 +1,664 @@
+	}
+
+	memset(&dma->blk_cfg, 0, sizeof(dma->blk_cfg));
+	dma->blk_cfg.source_address = (uintptr_t)&((IADC_TypeDef *)config->base)->SCANFIFODATA;
```
yes not needed removed.

Martinhoff-maker commented on drivers/adc/adc_silabs_iadc.c:
```diff
@@ -0,0 +1,664 @@
+	struct iadc_data *data = CONTAINER_OF(ctx, struct iadc_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
```
It now works with both moving pointer

jerome-pouiller reviewed (commented):
Only cosmetics comments. Good job!

Martinhoff-maker committed:
samples: adc_sequence: differential mode handling in ADC conversion

This commit updates the ADC sequence sample to improve handling of
differential mode readings. It modifies the conversion logic to
correctly interpret 16-bit/32-bit signed values and adjusts the
resolution accordingly. This ensures accurate millivolt conversion
for both differential and single-ended configurations.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

jerome-pouiller commented on drivers/adc/adc_silabs_iadc.c:
```diff
@@ -0,0 +1,688 @@
+			default:
+				LOG_ERR("Unsupported resolution %d", res);
+				return -EINVAL;
+			}
```
This function is rather long and this switch case appear twice in the function. Maybe we can factorize the code in one way or another?
I wonder if we could even leverage `ospl_table` to get `alignment` and `resolution`.

jerome-pouiller commented on drivers/adc/adc_silabs_iadc.c:
```diff
@@ -0,0 +1,688 @@
+		uint16_t resolution;
+	} ospl_table[] = {
+		{ 0, _IADC_CFG_OSRHS_HISPD2, _IADC_CFG_DIGAVG_AVG1,
+		 _IADC_SCANFIFOCFG_ALIGNMENT_RIGHT12, 12}, /* default */
```
space before brace.

jerome-pouiller commented on drivers/adc/adc_silabs_iadc.c:
```diff
@@ -0,0 +1,688 @@
+	} ospl_table[] = {
+		{ 0, _IADC_CFG_OSRHS_HISPD2, _IADC_CFG_DIGAVG_AVG1,
+		 _IADC_SCANFIFOCFG_ALIGNMENT_RIGHT12, 12}, /* default */
+		{ 1, _IADC_CFG_OSRHS_HISPD2, _IADC_CFG_DIGAVG_AVG1,
```
Nit: `wanted_oversampling` is unused and you can get this information using the table index (rule of the single root of trust). You could make the index explicit with:
```
    .[0] = {
        _IADC_CFG_OSRHS_HISPD2, _IADC_CFG_DIGAVG_AVG1,
        _IADC_SCANFIFOCFG_ALIGNMENT_RIGHT12, 12,
    },
```

asmellby commented on drivers/adc/adc_silabs_iadc.c:
```diff
@@ -0,0 +1,819 @@
+
+	for (int i = 0; i < ARRAY_SIZE(ospl_table); i++) {
+		if (ospl_table[i].resolution == res) {
+			data->resolution = ospl_table[i].resolution;
```
`data->resolution` seems like it's unused now, can it be removed from the data struct (and the ospl table)?

Martinhoff-maker replied:
yes you're right, since now we completely remove the possibility of applying the optimal resolution for a given oversampling (by that I mean just given an oversampling property), we can simplify a lot.

asmellby commented on drivers/adc/adc_silabs_iadc.c:
```diff
@@ -0,0 +1,819 @@
+	for (int i = 0; i < ARRAY_SIZE(ospl_table); i++) {
+		if (ospl_table[i].resolution == res) {
+			data->resolution = ospl_table[i].resolution;
+			data->alignment = ospl_table[i].alignment;
```
Since alignment is a function of resolution, could it be directly set in the switch on line 421 instead of in this loop? That would remove the need to keep a bunch of copies of it in the ospl table too.

Martinhoff-maker replied:
That's come from a wish to separate the input verification and the algo to find the best parameters but since we simplify a lot with no more res,etc... , yes.

Martinhoff-maker commented:
Need to wait for https://github.com/zephyrproject-rtos/zephyr/pull/96859 to be merged to have the CI green after a rebase.

JarmouniA reviewed (commented):
Please take a look at the discussion in https://github.com/zephyrproject-rtos/zephyr/pull/94585#issuecomment-3334665047 regarding all the samples' overlays/conf files added in this PR, it's not scalable at all and they have little added value of any.
Non-blocking

jhedberg commented:
@JarmouniA the points you raise about `platform_allow` require broader discussion within the project. PR review comments don't give the issue the attention it deserves, IMO. This was also a topic that was raised at the recent TSC F2F meeting a month ago. I believe one reason why these keep getting added is downstream on-HW CI, since not having the platforms listed means you have to explicitly force twister to use a given platform for testing, rather than e.g. requesting it to "test on all Silabs platforms".

Martinhoff-maker committed:
boards: silabs: add ADC io-channels on multiples boards

The goal to add the zephyr,user io-channels is to be able to run the
ADC samples without having an overlays for each boards.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker commented:
V5: 
- In order to be compliant with futur PR of @JarmouniA, remove all new sample overlays and instead introduce io-channels in some boards.dts
- Still a little bit confused about how we handle tests overlays since we have "differences" between SoC even if it's the "same" IP as Johan said.

jerome-pouiller commented on tests/drivers/adc/adc_api/testcase.yaml:
```diff
@@ -88,3 +88,32 @@ tests:
+      - xg24_rb4187c
+      - xg27_dk2602a
+      - bg29_rb4420a
+      - xg29_rb4412a
```
Note there are several revision of the IP (4 revisions I remember right). I agree it may not make sense to maintain one sample instance for each board. But I believe we need to cover all the boards in the tests (especially if you consider some CI are required to *run* these tests for the compliance with quality standards)

Don't misunderstand me, I agree with you. I would like to see the tests work out of the box without all the .overlay

Martinhoff-maker commented on doc/releases/migration-guide-4.3.rst:
```diff
@@ -61,6 +61,11 @@ Device Drivers and Devicetree
+ADC
+===
+
+* :dtcompatible:`silabs,gecko-iadc` has been replaced by :dtcompatible:`silabs,iadc`.
```
Since the new driver have a lot's of differences than the old gecko driver, I don't want to remove the old driver and old compatible for the moment. But then I think you're right, we might add "deprecated" for the old gecko iadc driver but how ?

Martinhoff-maker replied:
Thanks, i've added the deprecated symbol to the old iadc driver and updated the migration guide.

Martinhoff-maker commented on tests/drivers/adc/adc_api/testcase.yaml:
```diff
@@ -83,3 +83,32 @@ tests:
+      - EXTRA_CONF_FILE="overlay-dma-silabs.conf"
+      - CONFIG_ADC_ASYNC=y
+    depends_on:
+      - adc
```
removed

Martinhoff-maker commented on tests/drivers/adc/adc_api/testcase.yaml:
```diff
@@ -83,3 +83,32 @@ tests:
+    extra_args:
+      - EXTRA_CONF_FILE="overlay-dma-silabs.conf"
+    depends_on:
+      - adc
```
removed

Martinhoff-maker commented on dts/arm/silabs/xg29/xg29.dtsi:
```diff
@@ -535,7 +535,7 @@
 
 		adc0: adc@5a004000 {
-			compatible = "silabs,gecko-iadc";
+			compatible = "silabs,iadc";
```
So if I understand right, removed ```default=y``` in the old Kconfig and add:
```depends on (DT_HAS_SILABS_IADC_ENABLED || DT_HAS_SILABS_GECKO_IADC_ENABLED)``` in the the new Kconfig ? 
With that conf, the old binding will use the new driver but how can a user decide to go back to the old ? selecting the old Kconfig symbol will create a conflict. Or will the user needs to do:
```
CONFIG_ADC_SILABS_IADC=n
CONFIG_ADC_GECKO_IADC=y
```
and change the binding ?
It makes me thinking that I might prefer to add a binding than replace it like:
```compatible = "silabs,iadc", "silabs,gecko-iadc";```

Martinhoff-maker committed:
drivers: adc: silabs: removed the old IADC driver

iadc_gecko driver has been replaced by adc_silabs_iadc.c.
The old files and bindings are removed.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
drivers: adc: silabs: unify silabs adc kconfig

To reduce the number of Kconfig files and improve maintainability,
we are merging the three different Kconfig files from Silabs for ADC.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>
